include/seqan3/range/metafunction.hpp

```diff
@@ -54,78 +58,401 @@ namespace seqan3
  */
 
 // ----------------------------------------------------------------------------
+// uncvref
+// ----------------------------------------------------------------------------
 
-/*!\brief Whether a type has a `value_type` member and/or `ranges::value_type_t<t>` resolves [Value metafunction].
- * \tparam t The type to recurse on; must have `ranges::value_type_t<t>`
- * \hideinitializer
+/*!\brief Return the input type with references removed [Type metafunction].
+ * \tparam t The type to operate on.
+ */
+template <typename t>
+using unref_t = std::remove_reference_t<t>;
+
+// ----------------------------------------------------------------------------
+// uncvref
+// ----------------------------------------------------------------------------
+
+/*!\brief Return the input type with `const`, `volatile` and references removed [Type metafunction].
+ * \tparam t The type to operate on.
+ */
+template <typename t>
+using uncvref_t = std::remove_const_t<std::remove_volatile_t<std::remove_reference_t<t>>>;
+
+// ----------------------------------------------------------------------------
+// iterator
+// ----------------------------------------------------------------------------
+
+/*!\brief For a seqan3::range return its iterator type. [Type metafunction].
+ * \tparam t The type to operate on.
+ */
+template <typename t>
+using iterator_t = ranges::iterator_t<t>;
+
+// ----------------------------------------------------------------------------
+// sentinel
+// ----------------------------------------------------------------------------
+
+/*!\brief For a seqan3::range return its iterator type. [Type metafunction].
+ * \tparam t The type to operate on.
+ */
+template <typename t>
+using sentinel_t = ranges::sentinel_t<t>;
+
+// ----------------------------------------------------------------------------
+// value_type
+// ----------------------------------------------------------------------------
+
+/*!\brief Type metafunction that returns the `value_type` of another type [metafunction base template].
+ * \tparam t The type you wish to query.
+ *
+ * \details
+ *
+ * This is the base template, it needs to be overloaded.
+ *
+ * There is a shortcut for this metafunction: seqan3::value_type_t
+ */
+template <typename t>
+struct value_type;
+
+/*!\brief Type metafunction shortcut for seqan3::value_type.
+ * \tparam t The type you wish to query.
+ */
+template <typename t>
+using value_type_t = typename value_type<t>::type;
+
+/*!\brief Type metafunction that returns the `value_type` of another type [specialisation for input iterators].
+ * \tparam t The type you wish to query; must satisfy seqan3::input_iterator_concept.
+ */
+template <input_iterator_concept t>
+struct value_type<t>
```

Comment:
I'm still against it. I would prefer if we have a metafunction that can give you the iterator_type and then uses `iterator_traits` or that we us `range_traits` that is the same as `iterator_traits` for iterators, and defines the other stuff for ranges.